services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/model/MongoDbBasedCredentialsDao.java

```diff
@@ -0,0 +1,460 @@
+
+    /**
+     * Creates a new DAO.
+     *
+     * @param dbConfig The Mongo DB configuration properties.
+     * @param collectionName The name of the collection that contains the data.
+     * @param vertx The vert.x instance to run on.
+     * @throws NullPointerException if any of the parameters are {@code null}.
+     */
+    public MongoDbBasedCredentialsDao(
+            final MongoDbConfigProperties dbConfig,
+            final String collectionName,
+            final Vertx vertx) {
+        super(dbConfig, collectionName, vertx, null);
+    }
+
+    /**
+     * Creates a new DAO.
+     *
+     * @param dbConfig The Mongo DB configuration properties.
+     * @param collectionName The name of the collection that contains the data.
+     * @param vertx The vert.x instance to run on.
+     * @param tracer The tracer to use for tracking the processing of requests.
+     * @throws NullPointerException if any of the parameters other than tracer are {@code null}.
+     */
+    public MongoDbBasedCredentialsDao(
+            final MongoDbConfigProperties dbConfig,
+            final String collectionName,
+            final Vertx vertx,
+            final Tracer tracer) {
+        super(dbConfig, collectionName, vertx, tracer);
+    }
+
+    /**
+     * Creates the indices in the MongoDB that can be used to make querying of data more efficient.
+     *
+     * @return A succeeded future if the indices have been created. Otherwise, a failed future.
+     */
+    public Future<Void> createIndices() {
+
+        final Promise<Void> result = Promise.promise();
+
+        if (creatingIndices.compareAndSet(false, true)) {
+
+            // create unique index on tenant and device ID
+            createIndex(
+                    new JsonObject()
+                            .put(RegistryManagementConstants.FIELD_PAYLOAD_TENANT_ID, 1)
+                            .put(RegistryManagementConstants.FIELD_PAYLOAD_DEVICE_ID, 1),
+                    new IndexOptions().unique(true))
+                // create unique index on tenant, auth ID and type
+                .compose(ok -> createIndex(
+                        new JsonObject()
+                                .put(RegistryManagementConstants.FIELD_PAYLOAD_TENANT_ID, 1)
+                                .put(KEY_AUTH_ID, 1)
+                                .put(KEY_CREDENTIALS_TYPE, 1),
+                        new IndexOptions().unique(true)
+                                .partialFilterExpression(new JsonObject()
+                                        .put(KEY_AUTH_ID, new JsonObject().put("$exists", true))
+                                        .put(KEY_CREDENTIALS_TYPE, new JsonObject().put("$exists", true)))))
+                .onSuccess(ok -> indicesCreated.set(true))
+                .onComplete(r -> {
+                    creatingIndices.set(false);
+                    result.handle(r);
+                });
+        } else {
+            LOG.debug("already trying to create indices");
```

Comment:
actually I think that we should adapt the other DAOs to use the approach used here. If the indices are already being created then we simply let the client wait for its completion based on the existing future instead of telling it that the indices cannot be created. Doesn't that make more sense?

Reply:
It makes sense to wait instead of throwing `ConcurrentModificationException`.

Reply:
Ok, then I will adapt the other DAOs accordingly ...